### PR TITLE
foureight84 keymap for sofle

### PR DESCRIPTION
**Pull request by foureight84**
## Description

Add foureight84 keymap to the sofle keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
You need to rebase your branch, as you need to update the chibiOS libraries, so they're not being reverted here.  Rebasing sometimes will fix this. Otherwise, you'd want to do something like this to fix them:
```sh
git checkout master
git submodule foreach 'git log -n1 --oneline'
# remember the first abcdef1234 on the submodules that have an issue
git checkout @@your-branch@@
# for each of the problematic repo's:
cd lib/@@name@@
git checkout abcdef1234
```


Additionally, please review the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and verify that you have followed the checklist for the type of PR this is.  I've unchecked the "I have read the PR Checklist documentation" option, as there are changes listed there that have not been applied.

Eg, all c/h files need to have GPL 2+ compatible license headers. 

Additionally, you may want to change your oled wpm code, as s(n)printf uses ~1.5kb of firmware space.
https://docs.qmk.fm/#/squeezing_avr?id=oled-tweaks

**Comment by foureight84** [comment]
Gotcha. Thanks for this information.

File: keyboards/sofle/keymaps/foureight84/config.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+#define SPLIT_WPM_ENABLE
+#define SPLIT_OLED_ENABLE
+
+#ifndef NO_DEBUG
+#define NO_DEBUG
+#endif // !NO_DEBUG
+#if !defined(NO_PRINT) && !defined(CONSOLE_ENABLE)
+#define NO_PRINT
+#endif // !NO_PRINT
```
Comment (drashna) [inline keyboards/sofle/keymaps/foureight84/config.h:35]: This isn't needed.
```suggestion
```

File: keyboards/sofle/keymaps/foureight84/README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+For more information on integrating the Pimoroni Trackball, please see: https://github.com/foureight84/sofle-keyboard-pimoroni
```
Comment (drashna) [inline keyboards/sofle/keymaps/foureight84/README.md:1]: Please rename this file to be lower case.

File: keyboards/sofle/keymaps/foureight84/keymap.c [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+    KC_RAISE,
+    KC_ADJUST,
+    KC_PRVWD,
+    KC_NXTWD,
+    KC_LSTRT,
+    KC_LEND,
+    KC_DLINE,
+    KC_MOUSE,
+    KC_MOUSE_BTN1,
+    KC_MOUSE_BTN2
```
Comment (drashna) [inline keyboards/sofle/keymaps/foureight84/keymap.c:49]:
```suggestion
```

File: keyboards/sofle/keymaps/foureight84/keymap.c [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+        memcpy(&last_mouse_report, &mouse_report, sizeof(mouse_report));
+    }
+
+    return smooth_mouse_movement(mouse_report);
+}
+
+void keyboard_post_init_kb(void) {
+
+    keyboard_post_init_user();
+}
```
Comment (drashna) [inline keyboards/sofle/keymaps/foureight84/keymap.c:627]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/sofle/keymaps/foureight84/keymap.c [hunk 5]
```diff
@@ -0,0 +1,620 @@
...
+ */
+
+#include QMK_KEYBOARD_H
+#include <string.h>
+#include "lib/lib8tion/lib8tion.h"
+
+#ifdef OLED_ENABLE
+#include <stdio.h>
+
+#endif
```
Comment (drashna) [inline keyboards/sofle/keymaps/foureight84/keymap.c:24]:
```suggestion
```

**Review by drashna — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**